### ISSUE TITLE
"Deep props": is a context-like API without interior mutability possible?

### ISSUE THREAD

**Issue by rMazeiks** [DioxusLabs/dioxus#497]
I would like a way to provide props to a component's (possibly indirect) descendants. `use_context`(`_provider`) are very similar to what I want, but I want it to be more similar to props and more ergonomic (and I don't want children to be able to update the state), so I'm trying to write my own hooks from scratch.

The ideal API would look like this:

- `use_deep_prop_provider<T>(&cx, initializer: FnOnce() -> T) -> DeepPropProvider<T>`. The handle's API would be pretty much like `use_state`, but also provide the value to descendants.

- `use_deep_prop<T>(&cx) -> Option<&T>` would access a value if provided in an ancestor. Note: not in some wrapper, I just want a plain `&T`.

Intuitively, I feel like this should be possible – it's just like props, except deeper. One hook puts `T` in (through initializer and `set_state`), other hooks take `&T` out (ofc we can't  take `T` out bc there might be multiple users).

However, I don't think the current `ScopeState` allows this. Sure, we can take a value from the context and register "listener scopes" to be re-rendered when the value changes. But this lookup is expensive, and we don't want to do it on every render. So the current approach of `use_context` is to wrap stuff in `Rc<RefCell<>>` and re-use it on subsequent renders.

But then, we have an interior mutability thingy (`RefCell`), which intuitively seems not necessary (we don't need it for regular props). Also, it's not as ergonomic – we can't return a simple `&T`, since it must not live longer than the RefCell lock.

Is there a way we could borrow something that an ancestor has provided, and re-render every time it changes, but without the interior mutability?

I can't quite wrap my head around how this would be implemented. I'm guessing something needs to be added framework-level. Thoughts?

tldr: I want something like `use_context` but more similar to props.

---

btw, is there a reason why `consume_context` clones values before returning? Since it is guaranteed that parent scopes live longer than children, i would expect that it is safe to return a reference.

**Comment by jkelleyrtp** [comment]
Hmmmm. To provide &T without cloning would require the use of raw pointers. In theory you could provide &mut T so long as you only provide it to one component in your tree. 

Essentially you'd pin+box the value and then snatch it from up the tree. For what it's worth, *const T **is** clone so you could build a hook that uses the existing context API after boxing or hook allocating the original value. Of course, like borrowed props, it requires unsafe, but in this case, it would be sound.

**Comment by rMazeiks** [comment]
> Essentially you'd pin+box the value and then snatch it from up the tree. For what it's worth, *const T is clone so you could build a hook that uses the existing context API after boxing or hook allocating the original value. Of course, like borrowed props, it requires unsafe, but in this case, it would be sound.

Hmm, I'm confused how updating the value would work in this scenario. So you would still register `ScopeId`s to re-render? Would the dependents have to lookup the context again (slow)? Or would they still use the same pointer (surely that's unsound (?))?

**Comment by jkelleyrtp** [comment]
> > Essentially you'd pin+box the value and then snatch it from up the tree. For what it's worth, *const T is clone so you could build a hook that uses the existing context API after boxing or hook allocating the original value. Of course, like borrowed props, it requires unsafe, but in this case, it would be sound.
> 
> Hmm, I'm confused how updating the value would work in this scenario. So you would still register `ScopeId`s to re-render? ~Would the dependents have to lookup the context again (slow)?~ Or would they still use the same pointer ~(surely that's unsound (?))~?
> 
> edit: hmm pins

In almost every scenario dependents would need to lookup context. The use_context hook caches the value using the `clone` that you've noticed so it doesn't need to look up values on every render.

As for swapping state - you could always check if a new value is available when rendering the component. However, if you want to regenerate other components that are using the same value then you need a subscriber list.

You also need to be careful - if the value is swapped then it can't be dropped unless you guarantee that subscribers are dropped immediately after. In this case you'd need some way (like Rc) to know if/when a value should be dropped.

**Comment by rMazeiks** [comment]
After more thought, I've come to the conclusion that my requirements are inherently unsafe.

1. We want an ancestor component to have a handle that provides mutable access to the value (through a lock).
2. This handle needs to be cloned & passed around in handlers and such
3. We want a simple `&` reference to the value in children components

Unfortunately, 2. also means that we can't prevent the `UseContextProvider` being passed down through props (or can we?*). This means that, if my desired API were to be implemented, a child could get access to a `&` reference (through `use_context`) and a `UseContextProvider` (through props) providing mutable access, at the same time. This is Bad.

This means that my `use_context_provider` hook would need to be `unsafe` – the user must promise not to pass the handle down to children components. (`use_context` can be safe).

(rename: I ended up naming these `use_context` & -`_provider`, as I intend to use it as a replacement for the Dioxus hooks)

It feels weird to have an `unsafe` hook, but it could still be useful.

---

I did come up with an implementation like that. (I'm not sure how you meant to use `Pin` is in this situation, though – I didn't find it necessary; hopefully I didn't miss anything)

I found that I can simply share a `Rc<RefCell>`, and call the unsafe `try_borrow_unguarded` to get a `&` reference. I can guarantee that nobody will mutate the `RefCell` for the duration of the borrow, since the only way I allow to mutate it is through the provider handle, which must not be passed down to children (this is the `unsafe` contract of the context provider hook). Therefore, for the duration of the child render function (the lifetime of the `&`), there will be no mutation.

Is this reasoning sound? Are there edge cases I haven't thought of?

[Here's my code](https://gitlab.com/dawn_app/better-dioxus-hooks/-/blob/main/src/context.rs) for such an API. It [appears to work](https://gitlab.com/dawn_app/better-dioxus-hooks/-/blob/main/examples/context.rs)

---

Alternative: 

Instead of putting the unsafe in the `_provider`, we could put the burden on the `use_context_unsafe` user (and provide a `use_context` that would work with the lock as usual).

Pro: This would let the user decide if they want to use the (faster, more ergonomic) unsafe version, and uphold safety requirements manually, or the standard, runtime-lock version.

Cons: The safety requirement is a bit strange – We must ask to ensure that the context handle hasn't been passed down by a parent component, which the child really has no control over... I guess this would make the entire component unsafe to use? Also, the unsafe would spread to every consumer of the context, rather than having it in one place, the provider.

---

Anyway, I realize that, since this API fundamentally requires unsafe user code, it probably shouldn't be included in Dioxus, but possibly a 3rd party library. Feel free to close this issue. However, I would still appreciate your thoughts on the above, especially if you can think of an alternative I've overlooked. Thanks! 

*I thought about creating some auto-trait `PropSafe` that would be implemented for all types except those that must not be present in props. However, there's many other ways to get a value to a child (e.g. global state). | Can we somehow ensure at compile time that the handle only be used in the component and its handlers? Perhaps some cursed lifetime magic involving requiring a reference to `cx` or something? Just a thought, idk.
